src/util/type.cpp

```diff
@@ -34,3 +36,41 @@ bool is_number(const typet &type)
          id==ID_floatbv ||
          id==ID_fixedbv;
 }
+
+bool is_constant_or_has_constant_components(
+  const typet &type, const namespacet &ns)
+{
+  if (type.get_bool(ID_C_constant)) return true;
+
+  if (type.id() == ID_symbol) {
+    const auto &subtype = ns.follow(type);
+    if (subtype.id() == ID_struct || subtype.id() == ID_union) {
+      const auto &struct_union_type = to_struct_union_type(subtype);
+      for (const auto &component : struct_union_type.components())
+      {
+        if (component.type().id() == ID_pointer)
+          return component.type().get_bool(ID_C_constant);
+        if (component.type().get_bool(ID_C_constant))
+             return true;
+      }
+    }
+  }
+
+  if (type.has_subtype()) {
+    const auto &subtype = type.subtype();
+    if (subtype.id() == ID_symbol) {
+      const auto &subtype_id = ns.follow(to_symbol_type(subtype));
+      if (subtype_id.id() == ID_struct || subtype_id.id() == ID_union) {
+        const auto &struct_union_type = to_struct_union_type(subtype_id);
+        for (const auto &component : struct_union_type.components()) {
+            if (component.type().id() == ID_pointer)
+              return component.type().get_bool(ID_C_constant);
+            if (component.type().get_bool(ID_C_constant))
+              return true;
+        }
```

Comment:
This feels like a large amount of code duplication... and because of the code duplication it makes me wonder a little if the logic is fully right, but I'm not really sure. It looks to me like the first if block is identical to the `if (subtype.id() == ID_symbol) {}` block in the second block - so I'd start by factoring that into a helper function. You may then find that what you are left with in this function can be further restructured and simplified...